pkg/loadtester/agentloadtester.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2021-2024 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loadtester
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/charmbracelet/lipgloss"
+	"github.com/livekit/livekit-cli/v2/pkg/util"
+	"github.com/livekit/protocol/auth"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/utils"
+	lksdk "github.com/livekit/server-sdk-go/v2"
+	"github.com/pion/webrtc/v4"
+	"github.com/pion/webrtc/v4/pkg/media"
+	"go.uber.org/atomic"
+	"golang.org/x/sync/errgroup"
+)
+
+type LoadTestRoomStats struct {
+	agentDispatchedAt    time.Time
+	agentJoinedAt        time.Time
+	agentJoined          bool
+	agentTrackSubscribed bool
+	echoTrackPublished   bool
+	meetLink             string
+}
+
+type LoadTestRoom struct {
+	params           AgentLoadTestParams
+	room             *lksdk.Room
+	firstParticipant *lksdk.RemoteParticipant
+	echoTrack        *lksdk.LocalTrack
+	running          atomic.Bool
+	stats            LoadTestRoomStats
+}
+
+type AgentLoadTester struct {
+	params    AgentLoadTestParams
+	testRooms map[string]*LoadTestRoom
+	lock      sync.Mutex
+}
+
+func NewAgentLoadTester(params AgentLoadTestParams) *AgentLoadTester {
+	return &AgentLoadTester{
+		params:    params,
+		testRooms: make(map[string]*LoadTestRoom),
+	}
+}
+
+func NewLoadTestRoom(params AgentLoadTestParams) *LoadTestRoom {
+	return &LoadTestRoom{
+		params: params,
+		stats:  LoadTestRoomStats{},
+	}
+}
+
+func (t *AgentLoadTester) Start(ctx context.Context) error {
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	for i := 0; i < t.params.Rooms; i++ {
+		roomName := utils.NewGuid(fmt.Sprintf("room-%d-", i))
+
+		loadTestRoom := &LoadTestRoom{
+			params: t.params,
+		}
+
+		t.lock.Lock()
+		t.testRooms[roomName] = loadTestRoom
+		t.lock.Unlock()
+
+		group.Go(func() error {
+			if err := loadTestRoom.start(roomName); err != nil {
+				log.Printf("Failed to connect to room %s: %v", roomName, err)
+				return err
+			}
+
+			_, err := loadTestRoom.publishEchoTrack()
+			if err != nil {
+				log.Printf("Failed to publish echo track to room %s: %v", roomName, err)
+				loadTestRoom.stop()
+				return err
+			}
+			loadTestRoom.stats.echoTrackPublished = true
+
+			err = loadTestRoom.dispatchAgent()
```

Comment:
Curious why agent needs to be manually dispatched to the room? Won't this happen automatically?